engine/servermaster/openapi.go

```diff
@@ -0,0 +1,201 @@
+// @Tags jobs
+// @Accept json
+// @Produce json
+// @Param tenant query string false "tenant id"
+// @Param project query string false "project id"
+// @Success 200
+// @Failure 400,500
+// @Router /api/v1/jobs [get]
+func (o *OpenAPI) ListJobs(c *gin.Context) {
+	tenantID := c.Query(apiOpVarTenantID)
+	projectID := c.Query(apiOpVarProjectID)
+	_, _ = tenantID, projectID
+	c.Status(http.StatusNotImplemented)
+}
+
+// SubmitJob submits a new job.
+// @Summary Submit a job
+// @Description submits a new job
+// @Tags jobs
+// @Accept json
+// @Produce json
+// @Success 202
+// @Failure 400,500
+// @Router	/api/v1/jobs [post]
+func (o *OpenAPI) SubmitJob(c *gin.Context) {
+	c.Status(http.StatusNotImplemented)
+}
+
+// QueryJob queries a detail information of a job.
+// @Summary Query a job
+// @Description query a detail information of a job
+// @Tags jobs
+// @Accept json
+// @Produce json
+// @Param job_id  path  string  true  "job id"
+// @Success 200
+// @Failure 400,500
+// @Router /api/v1/jobs/{job_id} [get]
+func (o *OpenAPI) QueryJob(c *gin.Context) {
+	jobID := c.Param(apiOpVarJobID)
+	_ = jobID
+	c.Status(http.StatusNotImplemented)
+}
+
+// PauseJob pauses a job.
+// @Summary Pause a job
+// @Description pause a job
+// @Tags jobs
+// @Accept json
+// @Produce json
+// @Param job_id  path  string  true  "job id"
+// @Success 202
+// @Failure 400,500
+// @Router /api/v1/jobs/{job_id}/pause [post]
+func (o *OpenAPI) PauseJob(c *gin.Context) {
+	jobID := c.Param(apiOpVarJobID)
+	_ = jobID
+	c.Status(http.StatusNotImplemented)
+}
+
+// CancelJob cancels a job.
+// @Summary Cancel a job
+// @Description cancel a job
+// @Tags jobs
+// @Accept json
+// @Produce json
+// @Param job_id  path  string  true  "job id"
+// @Success 202
+// @Failure 400,500
+// @Router /api/v1/jobs/{job_id}/cancel [post]
+func (o *OpenAPI) CancelJob(c *gin.Context) {
+	jobID := c.Param(apiOpVarJobID)
+	_ = jobID
+	c.Status(http.StatusNotImplemented)
+}
+
+// ForwardJobMaster forwards api requests to job master.
+func (o *OpenAPI) ForwardJobMaster(c *gin.Context) {
+	jobID := c.Param(apiOpVarJobID)
+	tenantID := c.Query(apiOpVarTenantID)
+	projectID := c.Query(apiOpVarProjectID)
+	_, _, _ = jobID, tenantID, projectID
+
+	ctx := c.Request.Context()
+	resp, err := o.server.QueryJob(ctx, &pb.QueryJobRequest{JobId: jobID})
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	if resp.Err != nil {
+		if resp.Err.Code == pb.ErrorCode_UnKnownJob {
+			c.AbortWithStatus(http.StatusNotFound)
+		} else {
+			c.AbortWithError(http.StatusInternalServerError, errors.New(resp.Err.String()))
+		}
+		return
+	}
+	if resp.JobMasterInfo == nil {
+		c.AbortWithError(http.StatusInternalServerError, errors.New("couldn't find job master info"))
+		return
+	}
+	// TODO: check tenant id and project id.
+	executorID := model.ExecutorID(resp.JobMasterInfo.ExecutorId)
+	addr, ok := o.server.executorManager.GetAddr(executorID)
```

Comment:
You can move the `ExecutorInfoProvider` interface to a common package and modify it to let it provide the information you need, such as the address of executors.